rust/libceed/src/operator.rs

```diff
@@ -529,10 +746,205 @@ impl<'a> Operator<'a> {
                 v.into().to_raw(),
             )
         };
-        self.op_core.ceed.check_error(ierr)?;
+        self.op_core.check_error(ierr)?;
         Ok(self)
     }
 
+    /// Get a slice of Operator inputs
+    ///
+    /// ```
+    /// # use libceed::prelude::*;
+    /// # fn main() -> Result<()> {
+    /// # let ceed = libceed::Ceed::default_init();
+    /// let qf = ceed.q_function_interior_by_name("Mass1DBuild")?;
+    ///
+    /// // Operator field arguments
+    /// let ne = 3;
+    /// let q = 4 as usize;
+    /// let mut ind: Vec<i32> = vec![0; 2 * ne];
+    /// for i in 0..ne {
+    ///     ind[2 * i + 0] = i as i32;
+    ///     ind[2 * i + 1] = (i + 1) as i32;
+    /// }
+    /// let r = ceed.elem_restriction(ne, 2, 1, 1, ne + 1, MemType::Host, &ind)?;
+    /// let strides: [i32; 3] = [1, q as i32, q as i32];
+    /// let rq = ceed.strided_elem_restriction(ne, 2, 1, q * ne, strides)?;
+    ///
+    /// let b = ceed.basis_tensor_H1_Lagrange(1, 1, 2, q, QuadMode::Gauss)?;
+    ///
+    /// // Operator fields
+    /// let op = ceed
+    ///     .operator(&qf, QFunctionOpt::None, QFunctionOpt::None)?
+    ///     .field("dx", &r, &b, VectorOpt::Active)?
+    ///     .field("weights", ElemRestrictionOpt::None, &b, VectorOpt::None)?
+    ///     .field("qdata", &rq, BasisOpt::Collocated, VectorOpt::Active)?;
+    ///
+    /// let inputs = op.inputs()?;
+    ///
+    /// assert_eq!(inputs.len(), 2, "Incorrect inputs array");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn inputs(&self) -> crate::Result<&[crate::OperatorField]> {
+        // Get array of raw C pointers for inputs
+        let mut num_inputs = 0;
+        let mut inputs_ptr = std::ptr::null_mut();
+        let ierr = unsafe {
+            bind_ceed::CeedOperatorGetFields(
+                self.op_core.ptr,
+                &mut num_inputs,
+                &mut inputs_ptr,
+                std::ptr::null_mut() as *mut bind_ceed::CeedInt,
+                std::ptr::null_mut() as *mut *mut bind_ceed::CeedOperatorField,
+            )
+        };
+        self.op_core.check_error(ierr)?;
+        // Convert raw C pointers to fixed length slice
+        let inputs_slice = unsafe {
+            std::slice::from_raw_parts(
+                inputs_ptr as *const crate::OperatorField,
+                num_inputs as usize,
+            )
+        };
```

Comment:
I got it to work, but I think this is the part that is throwing me. How do we know that rustc and gccrs will continue to implement our  libCEED Rust structs in such a way that this will always work?

Reply:
It has nothing to do with C layout, just whether a newtype holding a single pointer has the same memory representation as the pointer it holds. I believe that is guaranteed. We could check that each pointer is non-null, but if pointers being non-null is already a contract enforced on the C side, then it would be redundant.